### DISCUSSION TITLE
How to receive a character over UART?

### DISCUSSION THREAD

**Discussion by Residualstress** [bitcraze/discussions#1285] [Q&A]
Hi, I noticed there is an example about [Send character to Crazyflie over UART](https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/other-examples/uart-send-char/). 

But I didn't find a corresponding example on the receiving end  (Crazyflie). 

For my project, it will send the location information of objects from an embedded system to Crazyflie.  Now, I could know the structure of the data including the location information. But I couldn't change it. That means I couldn't insert the position information into [CPXPacket](https://github.com/bitcraze/crazyflie-lib-python/blob/a77b4023867c27d814c6820373ec2a7d158a2ef5/cflib/cpx/__init__.py#L60). 

- What I want to do: decode the data passed from the embedded system on App layer.

- What I have done: I have tried to setup the Crazyflie under this tutorial [UART communication](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/). But when I send data that is not in CPXPacket format, the crazyflie will restart and report this [error](https://github.com/bitcraze/crazyflie-firmware/blob/9fe3099f8e2f9560d3dbb7d9e94c70f07edc7aee/src/modules/src/cpx/cpx_uart_transport.c#L164). 
  <img width="695" alt="8afb19ab04403d5a2bc122aa1d13acc" src="https://github.com/bitcraze/discussions/assets/92587824/984b8167-92c7-4bf3-8193-e6f392816466">

- What I want to ask:

 1. With the [UART Communication](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/)'s settings for Crazyflie, Crazyflie checks each incoming message for compliance with the CPXPacket format, right?
 2. If I want to decode the message on App layer by myself, e.g., receive a character, how should I set Crazyflie to avoid the [error](https://github.com/bitcraze/crazyflie-firmware/blob/9fe3099f8e2f9560d3dbb7d9e94c70f07edc7aee/src/modules/src/cpx/cpx_uart_transport.c#L164)  while enabling UART2?

Looking forward to your reply!

**Comment by gemenerik** [comment]
I think you should be able to send your information over UART2 (not UART1 !). The Crazyflie does not expect CPX packages here by default. You can also try completely disabling CPX (`CONFIG_ENABLE_CPX`)

As for the app layer: You can check out [this (pre-cpx) example](https://github.com/gemenerik/crazyflie-optical-flow-obstacle-avoidance-app/blob/main/app_obstacle_avoidance/src/obstacle_avoidance.c) and adapt it to read from UART2 instead of UART1.

For more reading checkout [this](https://github.com/orgs/bitcraze/discussions/1248) and [this](https://github.com/orgs/bitcraze/discussions/1272)  discussion.
